File: LeetCode/0091. Decode Ways/Solution.java
/**
 * 269 / 269 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 39.8 MB 
 */
class Solution {
    public int numDecodings(String s) {
        int len = s.length();
        int[] dp = new int[len + 1]; 
        dp[0] = 1; 
        for (int i = 1; i <= len; i++) {
            if (s.charAt(i - 1) != '0') {
                dp[i] += dp[i - 1]; 
            }
            if (i > 1 && s.charAt(i - 2) != '0') {
                int decode = (s.charAt(i - 2) - '0') * 10 + (s.charAt(i - 1) - '0');
                if (10 <= decode && decode <= 26) {
                    dp[i] += dp[i - 2]; 
                }
            }
        }
        return dp[len];
    }
}

File: LeetCode/0091. Decode Ways/Solution.kt
class Solution {
    fun numDecodings(s: String): Int {
        if (s.isEmpty() || s == "0") return 0
        val dp = Array(s.length + 1) { 0 }

        dp[0] = 1

        (1..s.length).forEach {
            if ('0' != s[it-1]) {
                dp[it] += dp[it - 1]
            }
            if (it > 1 && "${s[it-2]}${s[it-1]}" >= "10" && "${s[it-2]}${s[it-1]}" <= "26") {
                dp[it] += dp[it - 2]
            }
        }
        return dp[s.length]
    }
}

fun main(args: Array<String>) {
    val solution = Solution()
    solution.numDecodings("").let(::println)        //0
    solution.numDecodings("0").let(::println)       //0
    solution.numDecodings("1").let(::println)       //1
    solution.numDecodings("10").let(::println)      //1
    solution.numDecodings("12").let(::println)      //2
    solution.numDecodings("26").let(::println)      //2
    solution.numDecodings("30").let(::println)      //0
    solution.numDecodings("132").let(::println)     //2
    solution.numDecodings("2685").let(::println)    //2
    solution.numDecodings("123321").let(::println)  //6
}
